capellambse_context_diagrams/collectors/default.py

```diff
@@ -259,3 +267,91 @@ def port_context_collector(
             info.ports.append(port)
 
     return iter(ctx.values())
+
+
+def add_derived_components_and_interfaces(
+    diagram: context.ContextDiagram, data: _elkjs.ELKInputData
+) -> None:
+    """Add hidden Boxes and Exchanges to ``obj``'s context.
+
+    The derived exchanges are displayed with a dashed line.
+    """
+    if not (derivator := DERIVATORS.get(type(diagram.target))):
+        return
+
+    derivator(diagram, data)
+
+
+def derive_from_functions(
+    diagram: context.ContextDiagram, data: _elkjs.ELKInputData
+) -> None:
+    """Derive Components from allocated functions of the context target.
+
+    A Component, a ComponentExchange and two ComponentPorts are added
+    to ``data``. These elements are prefixed with ``Derived-`` to
+    receive special styling in the serialization step.
+    """
+    assert isinstance(diagram.target, cs.Component)
+    ports = []
+    for fnc in diagram.target.allocated_functions:
+        ports.extend(port_collector(fnc, diagram.type))
+
+    context_box_ids = {child["id"] for child in data["children"]}
+    components: dict[str, cs.Component] = {}
+    for port in ports:
+        for fex in port.exchanges:
+            if isinstance(port, fa.FunctionOutputPort):
+                attr = "target"
+            else:
+                attr = "source"
+
+            try:
+                derived_comp = getattr(fex, attr).owner.owner
+                if (
+                    derived_comp == diagram.target
+                    or derived_comp.uuid in context_box_ids
+                ):
+                    continue
+
+                if derived_comp.uuid not in components:
+                    components[derived_comp.uuid] = derived_comp
+            except AttributeError:
+                ...
+
+    # TODO: Even out derived interfaces on each side
+
+    centerbox = data["children"][0]
+    for i, (uuid, derived_component) in enumerate(components.items(), 1):
+        box = makers.make_box(
+            derived_component,
+            no_symbol=diagram.display_symbols_as_boxes,
+        )
+        class_ = type(derived_comp).__name__
+        box["id"] = f"__Derived-{class_}:{uuid}"
```

Comment:
Since I'm not that familiar with context diagrams and everything around them, I'm not sure if the following is helpful here, but I'll just drop it here for you to think about.

To increase maintainability and also to make it easier to see the different usages, it may be a good idea to turn the `Derived' prefix into a constant `DERIVED_PREFIX' and use it inside the code. This has the advantage that changing this prefix at some point will be really easy by just changing the value of the constant, by searching for the constant we can see all usages, and when extending the functionality we can just use this constant and not have to search for the specific value.